File: layer/AffineTransformLayer.py
# -*- coding: UTF-8 -*-
# caffe_root = 'caffe/'
# import sys
# sys.path.insert(0, caffe_root + 'python')

import caffe
import numpy as np


# Input: bottom[0]: Source images
#        bottom[1]: Affine parameters
#
# Output: top[0]: Transformed images

class AffineTransformLayer(caffe.Layer):

    def setup(self, bottom, top):
        if len(bottom) != 2:
            raise Exception("Need 2 inputs.")

        img_shape = bottom[0].shape
        # if img_shape[2] != img_shape[3]:
        #     raise Exception("Need 1:1 square images inputs.")

        self.out_img_height = img_shape[2]
        self.out_img_width = img_shape[3]

        self.in_img_height = img_shape[2]
        self.in_img_width = img_shape[3]

        # self.A = np.zeros((bottom[0].shape[0], 2, 2), float)
        # self.t = np.zeros((bottom[0].shape[0], 2), float)
        self.outPixels = np.zeros((bottom[0].shape[0], self.out_img_height * self.out_img_width, 2), float)

        self.dx = np.zeros((bottom[0].shape[0], self.out_img_height * self.out_img_width), float)
        self.dy = np.zeros((bottom[0].shape[0], self.out_img_height * self.out_img_width), float)
        self.pixel_minmin = np.zeros((bottom[0].shape[0], 1, 2), float)
        self.pixel_maxmin = np.zeros((bottom[0].shape[0], self.out_img_width, 2), float)
        self.pixel_minmax = np.zeros((bottom[0].shape[0], self.out_img_height, 2), float)
        self.pixel_maxmax = np.zeros((bottom[0].shape[0], self.out_img_height * self.out_img_width, 2), float)

    def reshape(self, bottom, top):
        top[0].reshape(bottom[0].shape[0], 1, self.out_img_height, self.out_img_width)

    def forward(self, bottom, top):
        for n in range(bottom[0].shape[0]):
            img = bottom[0].data[n]
            transform = bottom[1].data[n]
            A = np.zeros((2, 2))

            A[0, 0] = transform[0]
            A[0, 1] = transform[1]
            A[1, 0] = transform[2]
            A[1, 1] = transform[3]
            t = transform[4:6]

            # print A
            A = np.linalg.inv(A)
            t = np.dot(-t, A)

            # self.A[n, :] = A
            # self.t[n, :] = t

            pixels = [(x, y) for x in range(self.out_img_width) for y in range(self.out_img_height)]
            pixels = np.array(pixels, dtype=np.float32)

            outPixels = np.dot(pixels, A) + t

            outPixels[:, 0] = np.clip(outPixels[:, 0], 0, self.in_img_height - 2)
            outPixels[:, 1] = np.clip(outPixels[:, 1], 0, self.in_img_height - 2)

            self.outPixels[n, :] = outPixels

            outPixelsMinMin = outPixels.astype('int32')
            outPixelsMaxMin = outPixelsMinMin + [1, 0]
            outPixelsMinMax = outPixelsMinMin + [0, 1]
            outPixelsMaxMax = outPixelsMinMin + [1, 1]

            dx = outPixels[:, 0] - outPixelsMinMin[:, 0]
            dy = outPixels[:, 1] - outPixelsMinMin[:, 1]

            self.dx[n, :] = dx
            self.dy[n, :] = dy
            self.pixel_minmin[n, 0] = outPixelsMinMin[0]
            self.pixel_maxmin[n, :] = outPixelsMaxMin[:self.out_img_width*self.out_img_height:self.out_img_height]
            self.pixel_minmax[n, :] = outPixelsMinMax[:self.out_img_height]
            self.pixel_maxmax[n, :] = outPixelsMaxMax

            pixels = pixels.astype('int32')

            outImg = np.zeros((1, self.out_img_height, self.out_img_width), float)
            outImg[0, pixels[:, 1], pixels[:, 0]] += (1 - dx) * (1 - dy) * img[0, outPixelsMinMin[:, 1], outPixelsMinMin[:, 0]]
            outImg[0, pixels[:, 1], pixels[:, 0]] += dx * (1 - dy) * img[0, outPixelsMaxMin[:, 1], outPixelsMaxMin[:, 0]]
            outImg[0, pixels[:, 1], pixels[:, 0]] += (1 - dx) * dy * img[0, outPixelsMinMax[:, 1], outPixelsMinMax[:, 0]]
            outImg[0, pixels[:, 1], pixels[:, 0]] += dx * dy * img[0, outPixelsMaxMax[:, 1], outPixelsMaxMax[:, 0]]

            top[0].data[n, :] = outImg

    def backward(self, top, propagate_down, bottom):
        pass

File: layer/C++/affine_transform_layer.cpp
#include <vector>


#include "caffe/layers/affine_transform_layer.hpp"
#include "caffe/util/math_functions.hpp"

#define CLIP(x, l, h) ((x) < (l) ? (l) : ((x) > (h) ? (h) : (x)))  

namespace caffe{

template <typename Dtype>
void AffineTransformLayer<Dtype>::LayerSetUp(const vector<Blob<Dtype>*>& bottom, 
                                             const vector<Blob<Dtype>*>& top){
    CHECK_EQ(bottom.size(), 2)<< "Require 2 inputs.";
    num = bottom[0]->shape(0);
    height = bottom[0]->shape(2);
    weight = bottom[0]->shape(3);
    count = bottom[0]->count(1);
}

template <typename Dtype>
void AffineTransformLayer<Dtype>::Reshape(const vector<Blob<Dtype>*>& bottom,
                                          const vector<Blob<Dtype>*>& top){
    vector<int> shape = bottom[0]->shape();
    shape[1] = 1;
    top[0]->Reshape(shape);
}

template <typename Dtype>
void AffineTransformLayer<Dtype>::Forward_cpu(const vector<Blob<Dtype>*>& bottom,
                                              const vector<Blob<Dtype>*>& top){
    for (int n=0; n < num; n++){
        Dtype A[4];
        Dtype bias[2];
        caffe_copy<Dtype>(4, bottom[1]->cpu_data() + n * 6, A);
        caffe_copy<Dtype>(2, bottom[1]->cpu_data() + n*6+4, bias);      //A, bias

        Dtype A_inv[4];
        Dtype A_val = A[0] * A[3] - A[1] * A[2];
        CHECK_NE(A_val, (Dtype)0.)<<"Value of matrix A is 0!";
        A_inv[0] = A[3] / A_val;
        A_inv[1] = -A[1] / A_val;
        A_inv[2] = -A[2] / A_val;
        A_inv[3] = A[0] / A_val;
        Dtype bias1[2];
        caffe_cpu_gemm<Dtype>(CblasNoTrans,CblasNoTrans,1, 2, 2, (Dtype)1.,
           bias, A_inv, (Dtype)0., bias1);
        bias[0] = -bias1[0]; 
        bias[1] = -bias1[1];                           //A_inv, new bias

        Dtype pixels[weight * height][2] ={(Dtype)0.};
        Dtype outPixels[weight * height][2]={(Dtype)0.};   
        for (int y = 0; y < height; y++){
            for (int x = 0; x < weight; x++){
                pixels[y * weight + x][0] = (Dtype)x;
                pixels[y * weight + x][1] = (Dtype)y;

                Dtype outPixels1[2] = {(Dtype)0.};
                caffe_cpu_gemm<Dtype>(CblasNoTrans, CblasNoTrans, 1, 2, 2, (Dtype)1.,
                    pixels[y * weight + x], A_inv, (Dtype)0., outPixels1); 
                outPixels[y * weight + x][0] = CLIP(outPixels1[0] + bias[0], 0, weight -2);
                outPixels[y * weight + x][1] = CLIP(outPixels1[1] + bias[1], 0, height -2);     
            }
        }                                             // pixels  outPixels

        int outPixelsMinMin[weight * height][2];
        int outPixelsMaxMin[weight * height][2];
        int outPixelsMinMax[weight * height][2];
        int outPixelsMaxMax[weight * height][2];

        Dtype dx[weight * height];
        Dtype dy[weight * height];

        Dtype outImg[weight * height] = {(Dtype)0.};
        const Dtype* img = bottom[0]->cpu_data() + n * count;
        
        for (int y = 0; y < height; y++){
            for (int x = 0; x < weight; x++){
                outPixelsMinMin[y * weight + x][0] = (int) outPixels[y * weight + x][0];
                outPixelsMinMin[y * weight + x][1] = (int) outPixels[y * weight + x][1];
                outPixelsMaxMin[y * weight + x][0] = (int) outPixels[y * weight + x][0] + 1;
                outPixelsMaxMin[y * weight + x][1] = (int) outPixels[y * weight + x][1];
                outPixelsMinMax[y * weight + x][0] = (int) outPixels[y * weight + x][0];
                outPixelsMinMax[y * weight + x][1] = (int) outPixels[y * weight + x][1] + 1;
                outPixelsMaxMax[y * weight + x][0] = (int) outPixels[y * weight + x][0] + 1;
                outPixelsMaxMax[y * weight + x][1] = (int) outPixels[y * weight + x][1] + 1;

                dx[y * weight + x] = outPixels[y * weight + x][0] - outPixelsMinMin[y * weight + x][0];
                dy[y * weight + x] = outPixels[y * weight + x][1] - outPixelsMinMin[y * weight + x][1];

                outImg[y * weight + x] += (1 -dx[y * weight + x]) * (1 - dy[y * weight + x]) *
                    img[outPixelsMinMin[y * weight + x][1] * weight +outPixelsMinMin[y * weight +x][0]];
                outImg[y * weight + x] += dx[y * weight + x] * (1 - dy[y * weight + x]) *
                    img[outPixelsMaxMin[y * weight + x][1] * weight +outPixelsMaxMin[y * weight +x][0]];
                outImg[y * weight + x] += (1 -dx[y * weight + x]) * dy[y * weight + x] *
                    img[outPixelsMinMax[y * weight + x][1] * weight +outPixelsMinMax[y * weight +x][0]];
                outImg[y * weight + x] += dx[y * weight + x] * dy[y * weight + x] *
                    img[outPixelsMaxMax[y * weight + x][1] * weight +outPixelsMaxMax[y * weight +x][0]];
            }
        }                   //outPixelsMaxMax, outPixelsMaxMin, outPixelsMinMax,outPixelsMinMin, dx, dy


        caffe_copy<Dtype>(weight * height, outImg, top[0]->mutable_cpu_data() + n * weight * height);

    }
}

#ifdef CPU_ONLY
STUB_GPU(AffineTransformLayer);
#endif

INSTANTIATE_CLASS(AffineTransformLayer);
REGISTER_LAYER_CLASS(AffineTransform);

}
